layer_board.py

```diff
@@ -584,6 +592,42 @@ def discardLayersChanges(self, layerType='vector'):
         # Repopulate table
         self.populateLayerTable( layerType )
 
+    def setDataSource(self,layer,newSourceUri):
+        #method to apply a new datasource to a vector Layer
+        if "|" in newSourceUri:
+            newDatasourceType = newSourceUri.split("|")[0]
+            newUri = newSourceUri.split("|")[1].replace('\\','/')
+        else:
+            newDatasourceType = layer.dataProvider().name()
+            newUri = newSourceUri.replace('\\','/')
+        # read layer definition
+        XMLDocument = QDomDocument("style")
+        XMLMapLayers = QDomElement()
+        XMLMapLayers = XMLDocument.createElement("maplayers")
+        XMLMapLayer = QDomElement()
+        XMLMapLayer = XMLDocument.createElement("maplayer")
+        layer.writeLayerXML(XMLMapLayer,XMLDocument)
+        self.iface.setActiveLayer(layer)
+        
+        # probe new datasource to prevent layer issues
+        nlayer = QgsVectorLayer(newUri,"probe", newDatasourceType)
+        if not nlayer.isValid():
+            self.iface.messageBar().pushMessage("Error", "Incorrect source|uri string", level=QgsMessageBar.CRITICAL)
+            return
+        print "type", nlayer.geometryType(),layer.geometryType()
+        print "uri", newDatasourceType, newUri
+        if nlayer.geometryType() != layer.geometryType():
+            self.iface.messageBar().pushMessage("Error", "Geometry type mismatch", level=QgsMessageBar.CRITICAL)
+            return
```

Comment:
My concern is to be sure the user understand that the uri has not been changed, for example by adding a not in the log ?
It depends on when you do the controls on the uri. If we control the uri in the  onItemChanged and discard it here, or if we control when commiting changes (apply button).